props-core/src/main/java/sh/props/source/OnDemandSource.java

```diff
@@ -0,0 +1,168 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+package sh.props.source;
+
+import static java.lang.String.format;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiConsumer;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import sh.props.annotations.Nullable;
+
+/** Standard implementation of a {@link Source} capable of loading keys on-demand. */
+public abstract class OnDemandSource extends Source {
+  private static final Logger log = Logger.getLogger(OnDemandSource.class.getName());
+
+  private final Set<String> keys = ConcurrentHashMap.newKeySet();
+  private final Map<String, String> cache = new ConcurrentHashMap<>();
+
+  /**
+   * Loads the value for the given key.
+   *
+   * @param key a key that this source should retrieve
+   * @return the current value represented by the specified key
+   */
+  @Nullable
+  protected abstract String loadKey(String key);
+
+  /**
+   * Retrieves the list of keys which should be refreshed on-demand. Implementing subclasses are
+   * strongly encouraged to only retrieve these keys from the underlying store, but ultimately this
+   * is not enforced.
+   *
+   * @return a {@link Set} containing only the keys that were explicitly requested by a {@link
+   *     sh.props.Registry}
+   */
+  protected Set<String> getRegisteredKeys() {
+    return Collections.unmodifiableSet(keys);
+  }
+
+  /**
+   * Retrieves all keys registered with {@link #registerKey(String)} from the underlying store.
+   *
+   * @return a map containing the most recent
+   */
+  @Override
+  public Map<String, String> get() {
+    // refresh all registered keys
+    try {
+      // and wait until the request is processed by all sources
+      CompletableFuture.allOf(
+              getRegisteredKeys().stream()
+                  .map(this::retrieveKeyAsync)
+                  .toArray(CompletableFuture[]::new))
+          .join();
```

Comment:
refreshes all registered keys and returns the updated cache, on completion

Reply:
Future improvement: use a dedicated executor that can be configured by clients.  This impl. may not be efficient when dealing with a larger number of keys and may also cause concurrency issues due to the destination service's rate limiting.